In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

## Overview of the Data

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [3]:
%%sql
SELECT track_id, count(track_id) tracks_sold
    FROM invoice_line
    group by track_id;

Done.


[(1, 8),
 (2, 3),
 (3, 1),
 (4, 4),
 (5, 1),
 (6, 13),
 (7, 4),
 (8, 6),
 (9, 9),
 (10, 6),
 (11, 5),
 (12, 7),
 (13, 10),
 (14, 6),
 (15, 4),
 (16, 8),
 (17, 5),
 (18, 7),
 (19, 6),
 (20, 7),
 (21, 7),
 (22, 6),
 (23, 6),
 (24, 6),
 (25, 2),
 (26, 6),
 (27, 6),
 (28, 4),
 (29, 5),
 (30, 6),
 (31, 4),
 (32, 8),
 (33, 7),
 (34, 4),
 (35, 6),
 (36, 7),
 (37, 3),
 (38, 6),
 (39, 7),
 (40, 6),
 (41, 4),
 (42, 6),
 (43, 6),
 (44, 6),
 (45, 5),
 (46, 9),
 (47, 5),
 (48, 5),
 (49, 6),
 (50, 4),
 (51, 5),
 (52, 4),
 (53, 4),
 (54, 6),
 (55, 4),
 (56, 3),
 (57, 4),
 (58, 8),
 (59, 3),
 (60, 6),
 (61, 7),
 (62, 5),
 (63, 1),
 (64, 1),
 (65, 1),
 (66, 3),
 (67, 1),
 (68, 1),
 (69, 1),
 (70, 1),
 (71, 1),
 (72, 1),
 (73, 2),
 (74, 1),
 (75, 1),
 (76, 2),
 (77, 5),
 (78, 3),
 (79, 4),
 (80, 7),
 (81, 5),
 (82, 5),
 (83, 5),
 (84, 3),
 (85, 4),
 (86, 2),
 (87, 1),
 (88, 2),
 (89, 2),
 (90, 1),
 (91, 2),
 (92, 2),
 (93, 5),
 (94, 6),
 (95, 2),
 (96, 1),
 (97, 2),
 (98, 1),
 (100, 1),
 (102, 3),
 (103, 1),
 (105, 1),
 (108, 1),
 (109, 1),
 (110, 2),
 (126, 1),
 (128, 1),
 (130, 2),
 (131, 1),
 (132, 1),
 (134, 2),
 (135, 1),
 (137, 1),
 (138, 1),
 (139, 2),
 (140, 1),
 (142, 1),
 (146, 2),
 (147, 2),
 (149, 2),
 (150, 4),
 (151, 9),
 (152, 5),
 (153, 8),
 (154, 7),
 (155, 4),
 (156, 1),
 (157, 4),
 (158, 7),
 (160, 3),
 (161, 2),
 (162, 3),
 (163, 3),
 (164, 3),
 (165, 3),
 (172, 2),
 (183, 1),
 (184, 1),
 (186, 1),
 (188, 1),
 (194, 1),
 (195, 3),
 (196, 2),
 (197, 2),
 (198, 3),
 (199, 2),
 (200, 1),
 (201, 3),
 (202, 2),
 (203, 1),
 (204, 3),
 (206, 1),
 (207, 3),
 (208, 1),
 (212, 1),
 (219, 1),
 (222, 1),
 (224, 1),
 (225, 1),
 (226, 2),
 (227, 1),
 (228, 1),
 (229, 1),
 (233, 1),
 (237, 1),
 (261, 1),
 (268, 1),
 (298, 1),
 (301, 1),
 (337, 2),
 (340, 1),
 (341, 1),
 (342, 1),
 (344, 1),
 (346, 1),
 (348, 1),
 (349, 1),
 (350, 1),
 (383, 1),
 (392, 1),
 (402, 1),
 (409, 1),
 (410, 1),
 (411, 2),
 (413, 1),
 (415, 2),
 (419, 4),
 (420, 1),
 (421, 6),
 (422, 4),
 (423, 3),
 (424, 2),
 (425, 2),
 (426, 5),
 (427, 4),
 (428, 3),
 (429, 1),
 (430, 2),
 (431, 2),
 (432, 2),
 (433, 3),
 (434, 2),
 (435, 2),
 (436, 2),
 (437, 2),
 (438, 2),
 (439, 2),
 (440, 1),
 (441, 2),
 (442, 2),
 (443, 1),
 (444, 4),
 (445, 2),
 (446, 2),
 (447, 1),
 (448, 1),
 (449, 4),
 (450, 1),
 (451, 1),
 (452, 2),
 (453, 2),
 (454, 2),
 (455, 1),
 (459, 3),
 (461, 1),
 (465, 1),
 (468, 2),
 (469, 3),
 (470, 3),
 (472, 2),
 (473, 1),
 (474, 2),
 (475, 3),
 (476, 2),
 (477, 2),
 (478, 1),
 (479, 5),
 (480, 1),
 (481, 7),
 (482, 3),
 (483, 1),
 (484, 1),
 (485, 2),
 (486, 3),
 (487, 1),
 (488, 2),
 (492, 1),
 (496, 1),
 (510, 1),
 (514, 1),
 (517, 1),
 (524, 1),
 (528, 1),
 (529, 5),
 (530, 2),
 (531, 2),
 (532, 2),
 (533, 1),
 (534, 1),
 (535, 2),
 (536, 2),
 (537, 1),
 (538, 1),
 (539, 2),
 (540, 2),
 (541, 2),
 (542, 2),
 (544, 1),
 (545, 1),
 (546, 1),
 (548, 3),
 (553, 1),
 (571, 2),
 (572, 3),
 (573, 1),
 (574, 1),
 (577, 1),
 (579, 1),
 (580, 2),
 (585, 1),
 (587, 1),
 (592, 2),
 (593, 1),
 (594, 1),
 (597, 2),
 (599, 1),
 (600, 1),
 (602, 1),
 (604, 1),
 (605, 2),
 (608, 2),
 (609, 1),
 (610, 3),
 (611, 1),
 (612, 1),
 (614, 1),
 (615, 3),
 (619, 1),
 (620, 1),
 (622, 1),
 (675, 1),
 (676, 1),
 (681, 2),
 (682, 1),
 (684, 1),
 (686, 1),
 (687, 1),
 (688, 1),
 (689, 1),
 (690, 2),
 (691, 1),
 (692, 2),
 (693, 1),
 (695, 2),
 (696, 2),
 (697, 5),
 (698, 2),
 (699, 3),
 (700, 2),
 (701, 2),
 (702, 1),
 (703, 1),
 (704, 1),
 (705, 2),
 (706, 2),
 (707, 2),
 (708, 1),
 (709, 2),
 (710, 1),
 (711, 4),
 (712, 2),
 (713, 1),
 (714, 1),
 (719, 1),
 (739, 1),
 (740, 1),
 (748, 1),
 (749, 1),
 (750, 1),
 (751, 2),
 (755, 1),
 (757, 1),
 (768, 1),
 (771, 1),
 (773, 2),
 (778, 3),
 (781, 2),
 (784, 1),
 (787, 1),
 (788, 1),
 (794, 1),
 (795, 1),
 (800, 1),
 (802, 2),
 (803, 1),
 (804, 1),
 (807, 2),
 (808, 1),
 (809, 1),
 (810, 2),
 (811, 1),
 (812, 1),
 (813, 1),
 (814, 2),
 (815, 1),
 (818, 1),
 (819, 2),
 (822, 2),
 (826, 2),
 (827, 1),
 (828, 1),
 (829, 2),
 (830, 1),
 (

In [4]:
%%sql
SELECT 
    SUM(i.invoice_id)
             FROM customer c
          INNER JOIN invoice i ON i.customer_id = c.customer_id
        
        

Done.


[(188805,)]

In [5]:
%%sql
SELECT t.track_id, 
        g.name
FROM track t
INNER JOIN genre g ON g.genre_id = t.genre_id;

Done.


[(1, 'Rock'),
 (2, 'Rock'),
 (3, 'Rock'),
 (4, 'Rock'),
 (5, 'Rock'),
 (6, 'Rock'),
 (7, 'Rock'),
 (8, 'Rock'),
 (9, 'Rock'),
 (10, 'Rock'),
 (11, 'Rock'),
 (12, 'Rock'),
 (13, 'Rock'),
 (14, 'Rock'),
 (15, 'Rock'),
 (16, 'Rock'),
 (17, 'Rock'),
 (18, 'Rock'),
 (19, 'Rock'),
 (20, 'Rock'),
 (21, 'Rock'),
 (22, 'Rock'),
 (23, 'Rock'),
 (24, 'Rock'),
 (25, 'Rock'),
 (26, 'Rock'),
 (27, 'Rock'),
 (28, 'Rock'),
 (29, 'Rock'),
 (30, 'Rock'),
 (31, 'Rock'),
 (32, 'Rock'),
 (33, 'Rock'),
 (34, 'Rock'),
 (35, 'Rock'),
 (36, 'Rock'),
 (37, 'Rock'),
 (38, 'Rock'),
 (39, 'Rock'),
 (40, 'Rock'),
 (41, 'Rock'),
 (42, 'Rock'),
 (43, 'Rock'),
 (44, 'Rock'),
 (45, 'Rock'),
 (46, 'Rock'),
 (47, 'Rock'),
 (48, 'Rock'),
 (49, 'Rock'),
 (50, 'Rock'),
 (51, 'Rock'),
 (52, 'Rock'),
 (53, 'Rock'),
 (54, 'Rock'),
 (55, 'Rock'),
 (56, 'Rock'),
 (57, 'Rock'),
 (58, 'Rock'),
 (59, 'Rock'),
 (60, 'Rock'),
 (61, 'Rock'),
 (62, 'Rock'),
 (85, 'Rock'),
 (86, 'Rock'),
 (87, 'Rock'),
 (88, 'Rock'),
 (89, 'Rock'),
 (90, 'Rock'),
 (91, 'Rock'),
 (92, 'Rock'),
 (93, 'Rock'),
 (94, 'Rock'),
 (95, 'Rock'),
 (96, 'Rock'),
 (97, 'Rock'),
 (98, 'Rock'),
 (337, 'Rock'),
 (338, 'Rock'),
 (339, 'Rock'),
 (340, 'Rock'),
 (341, 'Rock'),
 (342, 'Rock'),
 (343, 'Rock'),
 (344, 'Rock'),
 (345, 'Rock'),
 (346, 'Rock'),
 (347, 'Rock'),
 (348, 'Rock'),
 (349, 'Rock'),
 (350, 'Rock'),
 (351, 'Rock'),
 (352, 'Rock'),
 (353, 'Rock'),
 (354, 'Rock'),
 (355, 'Rock'),
 (356, 'Rock'),
 (357, 'Rock'),
 (358, 'Rock'),
 (359, 'Rock'),
 (419, 'Rock'),
 (420, 'Rock'),
 (421, 'Rock'),
 (422, 'Rock'),
 (423, 'Rock'),
 (424, 'Rock'),
 (425, 'Rock'),
 (426, 'Rock'),
 (427, 'Rock'),
 (428, 'Rock'),
 (429, 'Rock'),
 (430, 'Rock'),
 (431, 'Rock'),
 (432, 'Rock'),
 (433, 'Rock'),
 (434, 'Rock'),
 (435, 'Rock'),
 (436, 'Rock'),
 (437, 'Rock'),
 (438, 'Rock'),
 (439, 'Rock'),
 (440, 'Rock'),
 (441, 'Rock'),
 (442, 'Rock'),
 (443, 'Rock'),
 (444, 'Rock'),
 (445, 'Rock'),
 (446, 'Rock'),
 (447, 'Rock'),
 (448, 'Rock'),
 (449, 'Rock'),
 (450, 'Rock'),
 (451, 'Rock'),
 (452, 'Rock'),
 (453, 'Rock'),
 (454, 'Rock'),
 (455, 'Rock'),
 (489, 'Rock'),
 (490, 'Rock'),
 (491, 'Rock'),
 (492, 'Rock'),
 (493, 'Rock'),
 (494, 'Rock'),
 (495, 'Rock'),
 (496, 'Rock'),
 (497, 'Rock'),
 (498, 'Rock'),
 (499, 'Rock'),
 (500, 'Rock'),
 (543, 'Rock'),
 (544, 'Rock'),
 (545, 'Rock'),
 (546, 'Rock'),
 (547, 'Rock'),
 (548, 'Rock'),
 (549, 'Rock'),
 (550, 'Rock'),
 (551, 'Rock'),
 (552, 'Rock'),
 (553, 'Rock'),
 (554, 'Rock'),
 (555, 'Rock'),
 (570, 'Rock'),
 (571, 'Rock'),
 (572, 'Rock'),
 (573, 'Rock'),
 (574, 'Rock'),
 (575, 'Rock'),
 (576, 'Rock'),
 (577, 'Rock'),
 (578, 'Rock'),
 (579, 'Rock'),
 (580, 'Rock'),
 (581, 'Rock'),
 (582, 'Rock'),
 (620, 'Rock'),
 (621, 'Rock'),
 (622, 'Rock'),
 (623, 'Rock'),
 (675, 'Rock'),
 (676, 'Rock'),
 (677, 'Rock'),
 (678, 'Rock'),
 (679, 'Rock'),
 (680, 'Rock'),
 (681, 'Rock'),
 (682, 'Rock'),
 (683, 'Rock'),
 (684, 'Rock'),
 (685, 'Rock'),
 (686, 'Rock'),
 (687, 'Rock'),
 (688, 'Rock'),
 (689, 'Rock'),
 (690, 'Rock'),
 (691, 'Rock'),
 (692, 'Rock'),
 (693, 'Rock'),
 (694, 'Rock'),
 (695, 'Rock'),
 (696, 'Rock'),
 (697, 'Rock'),
 (698, 'Rock'),
 (699, 'Rock'),
 (700, 'Rock'),
 (701, 'Rock'),
 (702, 'Rock'),
 (703, 'Rock'),
 (704, 'Rock'),
 (705, 'Rock'),
 (706, 'Rock'),
 (707, 'Rock'),
 (708, 'Rock'),
 (709, 'Rock'),
 (710, 'Rock'),
 (711, 'Rock'),
 (712, 'Rock'),
 (713, 'Rock'),
 (714, 'Rock'),
 (745, 'Rock'),
 (746, 'Rock'),
 (747, 'Rock'),
 (748, 'Rock'),
 (749, 'Rock'),
 (750, 'Rock'),
 (751, 'Rock'),
 (752, 'Rock'),
 (753, 'Rock'),
 (754, 'Rock'),
 (755, 'Rock'),
 (756, 'Rock'),
 (757, 'Rock'),
 (758, 'Rock'),
 (759, 'Rock'),
 (760, 'Rock'),
 (761, 'Rock'),
 (762, 'Rock'),
 (763, 'Rock'),
 (764, 'Rock'),
 (765, 'Rock'),
 (766, 'Rock'),
 (767, 'Rock'),
 (768, 'Rock'),
 (769, 'Rock'),
 (770, 'Rock'),
 (771, 'Rock'),
 (772, 'Rock'),
 (773, 'Rock'),
 (774, 'Rock'),
 (775, 'Rock'),
 (776, 'Rock'),
 (777, 'Rock'),
 (778, 'Rock'),
 (779, 'Rock'),
 (780, 'Rock'),
 (781, 'Rock'),
 (782

In [6]:
%%sql

WITH usa_customer AS (
        SELECT customer_id
        FROM customer
        WHERE country = "USA")
    track_genre AS(
        SELECT t.track_id, 
                g.name
        FROM track t
        INNER JOIN genre g ON g.genre_id = t.genre_id
            )
SELECT
        

(sqlite3.OperationalError) near "track_genre": syntax error
[SQL: WITH usa_customer AS (
        SELECT customer_id
        FROM customer
        WHERE country = "USA")
    track_genre AS(
        SELECT t.track_id, 
                g.name
        FROM track t
        INNER JOIN genre g ON g.genre_id = t.genre_id
            )
SELECT]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


## Selecting New Albums to Purchase

In [7]:
%%sql

WITH usa_tracks_sold AS
   (
    SELECT il.* FROM invoice_line il
    INNER JOIN invoice i on il.invoice_id = i.invoice_id
    INNER JOIN customer c on i.customer_id = c.customer_id
    WHERE c.country = "USA"
   )

SELECT
    g.name genre,
    count(uts.invoice_line_id) tracks_sold,
    cast(count(uts.invoice_line_id) AS FLOAT) / (
        SELECT COUNT(*) from usa_tracks_sold
    ) percentage_sold
FROM usa_tracks_sold uts
INNER JOIN track t on t.track_id = uts.track_id
INNER JOIN genre g on g.genre_id = t.genre_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10;

Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Latin', 22, 0.02093244529019981),
 ('Pop', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425)]

Based on the sales of tracks across different genres in the USA, we should purchase the new albums by the following artists:

Red Tone (Punk)
Slim Jim Bites (Blues)
Meteor and the Girls (Pop)
It's worth keeping in mind that combined, these three genres only make up only 17% of total sales, so we should be on the lookout for artists and albums from the 'rock' genre, which accounts for 53% of sales.

## Analyzing Employee Sales Performance

In [8]:
%%sql

WITH customer_buy AS (
    SELECT 
            i.customer_id,
            c.support_rep_id,
            SUM(i.total) total

    FROM invoice I
    INNER JOIN customer c ON c.customer_id = i.customer_id
    INNER JOIN employee e ON c.support_rep_id = e.employee_id
    GROUP BY 1,2 )

SELECT 
        e.first_name || " " || e.last_name name,
        e.title,
        SUM(cb.total) Total_sales
FROM employee e
INNER JOIN customer_buy cb ON cb.support_rep_id = e.employee_id
GROUP BY 1



Done.


[('Jane Peacock', 'Sales Support Agent', 1731.5099999999998),
 ('Margaret Park', 'Sales Support Agent', 1584.0000000000002),
 ('Steve Johnson', 'Sales Support Agent', 1393.92)]

While there is a 20% difference in sales between Jane (the top employee) and Steve (the bottom employee), the difference roughly corresponds with the differences in their hiring dates.

## Analyzing Sales by Country

In [9]:
%%sql



SELECT
    c.country,
    count(DISTINCT c.customer_id) AS total_number_of_customers,
    SUM(i.total) as total_sales,
    SUM(i.total)/count(c.customer_id) AS average_sales
    
FROM customer c
    
INNER JOIN invoice i ON c.customer_id = i.customer_id
GROUP BY 1
ORDER BY 3 DESC


Done.


[('USA', 13, 1040.4899999999998, 7.942671755725189),
 ('Canada', 8, 535.5900000000001, 7.047236842105265),
 ('Brazil', 5, 427.68000000000006, 7.0111475409836075),
 ('France', 5, 389.0699999999999, 7.781399999999998),
 ('Germany', 4, 334.62, 8.161463414634147),
 ('Czech Republic', 2, 273.24000000000007, 9.108000000000002),
 ('United Kingdom', 3, 245.52, 8.768571428571429),
 ('Portugal', 2, 185.13000000000002, 6.383793103448276),
 ('India', 2, 183.14999999999998, 8.72142857142857),
 ('Ireland', 1, 114.83999999999997, 8.833846153846151),
 ('Spain', 1, 98.01, 8.91),
 ('Chile', 1, 97.02000000000001, 7.463076923076924),
 ('Australia', 1, 81.18, 8.118),
 ('Finland', 1, 79.2, 7.2),
 ('Hungary', 1, 78.21, 7.821),
 ('Poland', 1, 76.22999999999999, 7.622999999999999),
 ('Sweden', 1, 75.24, 7.523999999999999),
 ('Norway', 1, 72.27000000000001, 8.030000000000001),
 ('Austria', 1, 69.3, 7.699999999999999),
 ('Netherlands', 1, 65.34, 6.534000000000001),
 ('Belgium', 1, 60.38999999999999, 8.627142857142855),
 ('Italy', 1, 50.49, 5.61),
 ('Argentina', 1, 39.6, 7.92),
 ('Denmark', 1, 37.61999999999999, 3.761999999999999)]

Based on the data, there may be opportunity in the following countries:

* Czech Republic
* United Kingdom
* India  
It's worth keeping in mind that because the amount of data from each of these countries is relatively low. Because of this, we should be cautious spending too much money on new marketing campaigns, as the sample size is not large enough to give us high confidence. A better approach would be to run small campaigns in these countries, collecting and analyzing the new customers to make sure that these trends hold with new customers.

## Albums vs Individual Tracks

In [11]:
%%sql

WITH invoice_first_track AS
    (
     SELECT
         il.invoice_id invoice_id,
         MIN(il.track_id) first_track_id
     FROM invoice_line il
     GROUP BY 1
    )

SELECT
    album_purchase,
    COUNT(invoice_id) number_of_invoices,
    CAST(count(invoice_id) AS FLOAT) / (
                                         SELECT COUNT(*) FROM invoice
                                      ) percent
FROM
    (
    SELECT
        ifs.*,
        CASE
            WHEN
                 (
                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 

                  EXCEPT 

                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id
                 ) IS NULL
             AND
                 (
                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 
                 ) IS NULL
             THEN "yes"
             ELSE "no"
         END AS "album_purchase"
     FROM invoice_first_track ifs
    )
GROUP BY album_purchase;

Done.


[('no', 500, 0.8143322475570033), ('yes', 114, 0.18566775244299674)]

Album purchases account for 18.6% of purchases. Based on this data, I would recommend against purchasing only select tracks from albums from record companies, since there is potential to lose one fifth of revenue.